src/amd/compiler/aco_lower_bool_phis.cpp

```diff
@@ -206,35 +207,32 @@ void lower_linear_bool_phi(Program *program, Block *block, aco_ptr<Instruction>&
 void lower_bool_phis(Program* program)
 {
    for (Block& block : program->blocks) {
-      std::vector<aco_ptr<Instruction>> instructions;
-      std::vector<aco_ptr<Instruction>> non_phi;
-      instructions.swap(block.instructions);
-      block.instructions.reserve(instructions.size());
-      unsigned i = 0;
-      for (; i < instructions.size(); i++)
-      {
-         aco_ptr<Instruction>& phi = instructions[i];
-         if (phi->opcode != aco_opcode::p_phi && phi->opcode != aco_opcode::p_linear_phi)
-            break;
-         if (phi->opcode == aco_opcode::p_phi && phi->definitions[0].regClass() == s2) {
-            non_phi.emplace_back(std::move(lower_divergent_bool_phi(program, &block, phi)));
-         } else if (phi->opcode == aco_opcode::p_linear_phi && phi->definitions[0].regClass() == s1) {
+      // find the first non linear phi
+      auto IsNonlinearPhi = [] (const aco_ptr<Instruction>& phi) -> bool {
+         return phi->opcode != aco_opcode::p_phi && phi->opcode != aco_opcode::p_linear_phi;
+      };
+      auto end = std::find_if(block.instructions.begin(), block.instructions.end(), IsNonlinearPhi);
+
+      // partition divergent phi's to the back
+      auto IsNotDivergentPhi = [] (const aco_ptr<Instruction>& phi) -> bool {
+         return !(phi->opcode == aco_opcode::p_phi && phi->definitions[0].regClass() == s2);
+      };
+      auto middle = std::stable_partition(block.instructions.begin(), end, IsNotDivergentPhi);
+
+      // lower the phi's
```

Comment:
/* lower the _boolean_ phis */
The idea or general problem we try to resolve here is that phis merge variables at join points, but bools are represented as 1bit per lane, i.e. one 64bit scalar register. For divergent control flow, the GPU executes both sides, and thus, would overwrite the previous result in the register instead of merging them (you can see how they get merged using AND/OR in lower_divergent_bool_phi()).